statsmodels/gam/tests/test_gam.py

```diff
@@ -0,0 +1,791 @@
+
+    s1 = np.dot(ccs.basis[:, ccs.mask[1]],
+                gam_res.params[ccs.mask[1]])
+    s1 -= s1.mean()  # TODO: Mean has to be removed
+
+    # plt.subplot(2, 1, 1)
+    # plt.plot(x[:, 0], s0, '.', label='s0')
+    # plt.plot(x[:, 0], s_mgcv[:, 0], '.', label='s0_mgcv')
+    # plt.legend(loc='best')
+    #
+    # plt.subplot(2, 1, 2)
+    # plt.plot(x[:, 1], s1, '.', label='s1_est')
+    # plt.plot(x[:, 1], s_mgcv[:, 1], '.', label='s1_mgcv')
+    # plt.legend(loc='best')
+    # plt.show()
+
+    assert_allclose(s0, s_mgcv[:, 0], atol=0.02)
+    assert_allclose(s1, s_mgcv[:, 1], atol=0.33)
+
+
+def test_multivariate_cubic_splines():
+    np.random.seed(0)
+    from statsmodels.gam.smooth_basis import CubicSplines
+
+    n = 500
+    x1 = np.linspace(-3, 3, n)
+    x2 = np.linspace(0, 1, n)**2
+
+    x = np.vstack([x1, x2]).T
+    y1 = np.sin(x1) / x1
+    y2 = x2 * x2
+    y0 = y1 + y2
+    # need small enough noise variance to get good estimate for this test
+    y = y0 + np.random.normal(0, .3 / 2, n)
+    y -= y.mean()
+    y0 -= y0.mean()
+
+    alphas = [1e-3, 1e-3]
+    cs = CubicSplines(x, df=[10, 10], constraints='center')
+
+    gam = GLMGam(y, exog=np.ones((n, 1)), smoother=cs, alpha=alphas)
+    gam_res = gam.fit(method='pirls')
+
+    y_est = gam_res.fittedvalues
+    y_est -= y_est.mean()
+
+    # cut the tails
+    index = list(range(50, n - 50))
+    y_est = y_est[index]
+    y0 = y0[index]
+    y = y[index]
+
+    # plt.plot(y_est, label='y est')
+    # plt.plot(y0, label='y0')
+    # plt.plot(y, '.', label='y')
+    # plt.legend(loc='best')
+    # plt.show()
+
+    assert_allclose(y_est, y0, atol=0.04)
+
+
+def test_glm_pirls_compatibility():
+    np.random.seed(0)
+
+    n = 500
+    x1 = np.linspace(-3, 3, n)
+    x2 = np.random.rand(n)
+
+    x = np.vstack([x1, x2]).T
+    y1 = np.sin(x1) / x1
+    y2 = x2 * x2
+    y0 = y1 + y2
+    y = y0 + np.random.normal(0, .3, n)
+    y -= y.mean()
+    y0 -= y0.mean()
+
+    # TODO: we have now alphas == alphas_glm
+    alphas = [5.75] * 2
+    alphas_glm = [1.2] * 2
+    # using constraints avoids singular exog.
+    cs = BSplines(x, df=[10, 10], degree=[3, 3], constraints='center')
+
+    gam_pirls = GLMGam(y, smoother=cs, alpha=alphas)
+    gam_glm = GLMGam(y, smoother=cs, alpha=alphas)
+
+    gam_res_glm = gam_glm.fit(method='nm', max_start_irls=0,
+                              disp=1, maxiter=20000, maxfun=10000)
+    gam_res_glm = gam_glm.fit(start_params=gam_res_glm.params,
+                              method='bfgs', max_start_irls=0,
+                              disp=1, maxiter=20000, maxfun=10000)
+    gam_res_pirls = gam_pirls.fit()
+
+    y_est_glm = np.dot(cs.basis, gam_res_glm.params)
+    y_est_glm -= y_est_glm.mean()
+    y_est_pirls = np.dot(cs.basis, gam_res_pirls.params)
+    y_est_pirls -= y_est_pirls.mean()
+
+    # plt.plot(y_est_pirls)
+    # plt.plot(y_est_glm)
+    # plt.plot(y, '.')
+    # plt.show()
```

Comment:
below there is a nice comment on similarly commented-out code explaining why its there